pkg/controller/rollout/rollout_releaseManager.go

```diff
@@ -27,8 +38,154 @@ type ReleaseManager interface {
+	if br.Spec.ReleasePlan.BatchPartition == nil &&
+		br.Status.Phase == v1beta1.RolloutPhaseCompleted {
+		klog.Infof("rollout(%s/%s) finalizing batchRelease(%s) done", c.Rollout.Namespace, c.Rollout.Name, util.DumpJSON(br.Status))
+		return false, nil
+	}
+
+	// If BatchPartition is nil, BatchRelease will directly resume workload via:
+	// - * set workload Paused = false if it needs;
+	// - * set workload Partition = null if it needs.
+	if br.Spec.ReleasePlan.BatchPartition == nil {
+		// - If checkReady is true, finalizing policy must be "WaitResume";
+		// - If checkReady is false, finalizing policy must be NOT "WaitResume";
+		// Otherwise, we should correct it.
+		switch br.Spec.ReleasePlan.FinalizingPolicy {
+		case v1beta1.WaitResumeFinalizingPolicyType:
+			if waitReady { // no need to patch again
+				return true, nil
+			}
+		default:
+			if !waitReady { // no need to patch again
+				return true, nil
+			}
+		}
+	}
+
+	// Correct finalizing policy.
+	policy := v1beta1.ImmediateFinalizingPolicyType
+	if waitReady {
+		policy = v1beta1.WaitResumeFinalizingPolicyType
+	}
+
+	// Patch BatchPartition and FinalizingPolicy, BatchPartition always patch null here.
+	body := fmt.Sprintf(`{"spec":{"releasePlan":{"batchPartition":null,"finalizingPolicy":"%s"}}}`, policy)
+	if err = cli.Patch(context.TODO(), br, client.RawPatch(types.MergePatchType, []byte(body))); err != nil {
+		return true, err
+	}
+	klog.Infof("rollout(%s/%s) patch batchRelease(%s) success", c.Rollout.Namespace, c.Rollout.Name, body)
+	return true, nil
+}
+
+func runBatchRelease(m ReleaseManager, rollout *v1beta1.Rollout, rolloutId string, batch int32, isRollback bool) (bool, *v1beta1.BatchRelease, error) {
+	cli := m.fetchClient()
+	batch = batch - 1
+	br, err := fetchBatchRelease(cli, rollout.Namespace, rollout.Name)
+	if errors.IsNotFound(err) {
+		// create new BatchRelease Crd
+		br = m.createBatchRelease(rollout, rolloutId, batch, isRollback)
+		if err = cli.Create(context.TODO(), br); err != nil && !errors.IsAlreadyExists(err) {
+			klog.Errorf("rollout(%s/%s) create BatchRelease failed: %s", rollout.Namespace, rollout.Name, err.Error())
+			return false, nil, err
+		}
+		klog.Infof("rollout(%s/%s) create BatchRelease(%s) success", rollout.Namespace, rollout.Name, util.DumpJSON(br))
+		return false, br, nil
+	} else if err != nil {
+		klog.Errorf("rollout(%s/%s) fetch BatchRelease failed: %s", rollout.Namespace, rollout.Name, err.Error())
+		return false, nil, err
+	}
+
+	// check whether batchRelease configuration is the latest
+	newBr := m.createBatchRelease(rollout, rolloutId, batch, isRollback)
+	if reflect.DeepEqual(br.Spec, newBr.Spec) && reflect.DeepEqual(br.Annotations, newBr.Annotations) {
+		klog.Infof("rollout(%s/%s) do batchRelease batch(%d) success", rollout.Namespace, rollout.Name, batch+1)
+		return true, br, nil
+	}
+	// update batchRelease to the latest version
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
plz avoid retry directly in the reconcile loop, one can always retry by workerqueue

Reply:
ok